File: riskfolio/external/cppfunctions.cpp
/*
 * Copyright (c) 2020-2022, Dany Cajas
 * All rights reserved.
 * This work is licensed under BSD 3-Clause "New" or "Revised" License.
 * License available at https://github.com/dcajasn/Riskfolio-Lib/blob/master/LICENSE.txt
 */

#include "cppfunctions.h"
#include <cstdio>

using namespace std;
using namespace arma;
using namespace std::chrono;

/**
 * Calculate duplication matrix of size "n" as shown
 * in Magnus, J. R., & Neudecker, H. (1980). The Elimination Matrix:
 * Some Lemmas and Applications. In SIAM Journal on Algebraic Discrete
 * Methods (Vol. 1, Issue 4, pp. 422–449). Society for Industrial &
 * Applied Mathematics (SIAM). https://doi.org/10.1137/0601049
 *
 * @n size of symmetric matrix.
 */
arma::Mat<double> duplication_matrix(const int &n) {
    int cols = (n*(n+1))/2;
    int col;
    arma::sp_mat out = sp_mat(n*n, cols);
    for (int j = 0; j < n; ++j) {
        for (int i = j; i < n; ++i) {
            arma::sp_mat u = sp_mat(1, cols);
            col = j*n+i-((j+1)*j)/2;
            u(0, col) = 1.0;
            arma::sp_mat T = sp_mat(n,n);
            T(i,j) = 1.0;
            T(j,i) = 1.0;
            out += arma::vectorise(T) * u;
        }
    }
    arma::Mat<double> D = arma::mat(out);
    return D;
}

/**
 * Calculate duplication elimination matrix of size "n" as shown
 * in Magnus, J. R., & Neudecker, H. (1980). The Elimination Matrix:
 * Some Lemmas and Applications. In SIAM Journal on Algebraic Discrete
 * Methods (Vol. 1, Issue 4, pp. 422–449). Society for Industrial &
 * Applied Mathematics (SIAM). https://doi.org/10.1137/0601049
 *
 * @n size of symmetric matrix.
 */
arma::Mat<double> duplication_elimination_matrix(const int &n) {
    int rows = (n*(n+1))/2;
    int row;
    arma::sp_mat out = sp_mat(rows, n*n);
    for (int j = 0; j < n; ++j) {
        arma::sp_mat e_j = sp_mat(1, n);
        e_j(0, j) = 1.0;
        for (int i = j; i < n; ++i) {
            arma::sp_mat u = sp_mat(rows, 1);
            row = j*n+i-((j+1)*j)/2;
            u(row, 0) = 1.0;
            arma::sp_mat e_i = sp_mat(1, n);
            e_i(0, i) = 1.0;
            out += arma::kron(u, arma::kron(e_j, e_i));
        }
    }
    arma::Mat<double> L = arma::mat(out);
    return L;
}

/**
 * Calculates summation matrix of size "n" as shown in Cajas, D. (2022).
 * Convex Optimization of Portfolio Kurtosis. In SSRN Electronic Journal.
 * Elsevier BV. https://doi.org/10.2139/ssrn.4202967
 *
 * @n size of symmetric matrix.
 */
arma::Mat<double> summation_matrix(const int &n) {
    arma::mat D;
    arma::mat L;
    D = duplication_matrix(n);
    L = duplication_elimination_matrix(n);
    arma::sp_mat D0;
    arma::sp_mat L0;
    arma::sp_mat S0;
    D0 = arma::sp_mat(D);
    L0= arma::sp_mat(L);
    S0 = D0.t() * D0 * L0;
    arma::Mat<double> S;
    S = arma::mat(S0);
    return S;
}

/**
 * Calculates kurtosis square matrix as shown in Cajas, D. (2022).
 * Convex Optimization of Portfolio Kurtosis. In SSRN Electronic Journal.
 * Elsevier BV. https://doi.org/10.2139/ssrn.4202967
 *
 * @Y matrix of assets returns.
 */
arma::Mat<double> kurtosis_matrix(const arma::Mat<double> &Y) {
    double T = Y.n_rows;
    arma::mat mu = arma::mean(Y);
    arma::mat mu_mat(arma::size(Y), fill::zeros);
    mu_mat.each_row() = mu; 
    arma::mat X = Y - mu_mat;
    arma::mat ones(1, X.n_cols, fill::ones);
    arma::Mat<double> M4 = arma::kron(ones, X) % arma::kron(X, ones);
    M4 = 1/T * M4.t() * M4;
    return M4;
}

/**
 * Calculates semi-kurtosis square matrix as shown in Cajas, D. (2022).
 * Convex Optimization of Portfolio Kurtosis. In SSRN Electronic Journal.
 * Elsevier BV. https://doi.org/10.2139/ssrn.4202967
 *
 * @Y matrix of assets returns.
 */
arma::Mat<double> semi_kurtosis_matrix(const arma::Mat<double> &Y) {
    double T = Y.n_rows;
    arma::mat mu = arma::mean(Y);
    arma::mat mu_mat(arma::size(Y), fill::zeros);
    arma::mat zeros_mat(arma::size(Y), fill::zeros);
    mu_mat.each_row() = mu;
    arma::mat X = Y - mu_mat;
    X = arma::min(X, zeros_mat);
    arma::mat ones(1, X.n_cols, fill::ones);
    arma::Mat<double> SM4 = arma::kron(ones, X) % arma::kron(X, ones);
    SM4 = 1/T * SM4.t() * SM4;
    return SM4;
}

File: riskfolio/external/cppfunctions.py
""""""  #

"""
Copyright (c) 2020-2026, Dany Cajas
All rights reserved.
This work is licensed under BSD 3-Clause "New" or "Revised" License.
License available at https://github.com/dcajasn/Riskfolio-Lib/blob/master/LICENSE.txt
"""

import numpy as np
import pandas as pd
from riskfolio.external.functions import *
from itertools import product

__all__ = [
    "duplication_matrix",
    "duplication_elimination_matrix",
    "duplication_summation_matrix",
    "commutation_matrix",
    "coskewness_matrix",
    "semi_coskewness_matrix",
    "cokurtosis_matrix",
    "semi_cokurtosis_matrix",
    "k_eigh",
    "d_corr",
    "d_corr_matrix",
    "residuals_coskewness_fm",
    "residuals_cokurtosis_fm",
]


def duplication_matrix(n: int, diag: bool = True):
    r"""
    Calculate duplication matrix of size "n" as shown in :cite:`d-Magnus1980`.

    Parameters
    ----------
    n : int
        Number of assets.

    Returns
    -------
    D: np.ndarray
        Duplication matrix
    """
    return cpp_duplication_matrix(n, diag)


def duplication_elimination_matrix(n: int, diag: bool = True):
    r"""
    Calculate duplication elimination matrix of size "n" as shown in :cite:`d-Magnus1980`.

    Parameters
    ----------
    n : int
        Number of assets.

    Returns
    -------
    L: np.ndarray
        Duplication matrix
    """
    return cpp_duplication_elimination_matrix(n, diag)


def duplication_summation_matrix(n: int, diag: bool = True):
    r"""
    Calculate duplication summation matrix of size "n" as shown in :cite:`d-Cajas4`.

    Parameters
    ----------
    n : int
        Number of assets.

    Returns
    -------
    S: np.ndarray
        Duplication summation matrix.
    """
    return cpp_duplication_summation_matrix(n, diag)


def commutation_matrix(T: int, n: int):
    r"""
    Calculate commutation matrix of size T x n.

    Parameters
    ----------
    T : int
        Number of rows.

    n : int
        Number of columns.

    Returns
    -------
    K: np.ndarray
        Duplication summation matrix.
    """
    return cpp_commutation_matrix(T, n)


def coskewness_matrix(Y: np.ndarray):
    r"""
    Calculates coskewness rectangular matrix as shown in :cite:`d-Cajas4`.

    Parameters
    ----------
    Y : ndarray or dataframe
        Returns series of shape n_sample x n_features.

    Returns
    -------
    M3 : ndarray
        The lower semi coskewness rectangular matrix.

    Raises
    ------
        ValueError when the value cannot be calculated.

    """
    flag = False
    if isinstance(Y, pd.DataFrame):
        assets = Y.columns.tolist()
        cols = list(product(assets, assets))
        cols = [str(y) + " - " + str(x) for x, y in cols]
        flag = True

    Y_ = np.array(Y, ndmin=2)
    M3 = cpp_coskewness_matrix(Y_, semi=False)

    if flag:
        M3 = pd.DataFrame(M3, index=assets, columns=cols)

    return M3


def semi_coskewness_matrix(Y: np.ndarray):
    r"""
    Calculates lower semi coskewness rectangular matrix as shown in :cite:`d-Cajas4`.

    Parameters
    ----------
    Y : ndarray or dataframe
        Returns series of shape n_samples x n_features.

    Returns
    -------
    s_M3 : ndarray
        The lower semi coskewness rectangular matrix.

    Raises
    ------
        ValueError when the value cannot be calculated.

    """
    flag = False
    if isinstance(Y, pd.DataFrame):
        assets = Y.columns.tolist()
        cols = list(product(assets, assets))
        cols = [str(y) + " - " + str(x) for x, y in cols]
        flag = True

    Y_ = np.array(Y, ndmin=2)
    s_M3 = cpp_coskewness_matrix(Y_, semi=True)

    if flag:
        s_M3 = pd.DataFrame(s_M3, index=assets, columns=cols)

    return s_M3


def cokurtosis_matrix(Y: np.ndarray):
    r"""
    Calculates cokurtosis square matrix as shown in :cite:`d-Cajas4`.

    Parameters
    ----------
    Y : ndarray or dataframe
        Returns series of shape n_samples x n_features.

    Returns
    -------
    S4 : ndarray
        The cokurtosis square matrix.

    Raises
    ------
        ValueError when the value cannot be calculated.

    """
    flag = False
    if isinstance(Y, pd.DataFrame):
        assets = Y.columns.tolist()
        cols = list(product(assets, assets))
        cols = [str(y) + " - " + str(x) for x, y in cols]
        flag = True

    Y_ = np.array(Y, ndmin=2)
    S4 = cpp_cokurtosis_matrix(Y_, semi=False)

    if flag:
        S4 = pd.DataFrame(S4, index=cols, columns=cols)

    return S4


def semi_cokurtosis_matrix(Y: np.ndarray):
    r"""
    Calculates lower semi cokurtosis square matrix as shown in :cite:`d-Cajas4`.

    Parameters
    ----------
    Y : ndarray or dataframe
        Returns series of shape n_sample x n_features.

    Returns
    -------
    s_S4 : ndarray
        The lower semi cokurtosis square matrix.

    Raises
    ------
        ValueError when the value cannot be calculated.

    """
    flag = False
    if isinstance(Y, pd.DataFrame):
        assets = Y.columns.tolist()
        cols = list(product(assets, assets))
        cols = [str(y) + " - " + str(x) for x, y in cols]
        flag = True

    Y_ = np.array(Y, ndmin=2)
    s_S4 = cpp_cokurtosis_matrix(Y_, semi=True)

    if flag:
        s_S4 = pd.DataFrame(s_S4, index=cols, columns=cols)

    return s_S4


def k_eigh(Y: np.ndarray, k: int):
    r"""
    Calculates lower semi cokurtosis square matrix as shown in :cite:`d-Cajas4`.

    Parameters
    ----------
    Y : ndarray or dataframe
        Returns series of shape n_sample x n_features.

    Returns
    -------
    s_S4 : ndarray
        The lower semi cokurtosis square matrix.

    Raises
    ------
        ValueError when the value cannot be calculated.
    """
    Y_ = np.array(Y, ndmin=2)
    eigvalues, eigvectors = cpp_k_eigh(Y_, k)

    return eigvalues, eigvectors


def d_corr(X: np.ndarray, Y: np.ndarray):
    r"""
    Calculates the distance correlation of X and Y.

    Parameters
    ----------
    X : ndarray or dataframe
        Returns series of shape n_sample x n_features.
    Y : ndarray or dataframe
        Returns series of shape n_sample x n_features.

    Returns
    -------
    value : float
        Distance correlation.

    Raises
    ------
        ValueError when the value cannot be calculated.
    """
    X_ = np.array(X, ndmin=2)
    Y_ = np.array(Y, ndmin=2)
    value = cpp_dcorr(X_, Y_)

    return value


def d_corr_matrix(Y: np.ndarray):
    r"""
    Calculates the distance correlation matrix of matrix of variables Y.

    Parameters
    ----------
    Y : ndarray or dataframe
        Returns series of shape n_sample x n_features.

    Returns
    -------
    value : float
        Distance correlation.

    Raises
    ------
        ValueError when the value cannot be calculated.

    """
    Y_ = np.array(Y, ndmin=2)
    value = cpp_dcorr_matrix(Y_)

    return value


def residuals_coskewness_fm(residuals: np.ndarray):
    r"""
    Calculates the coskewness tensor of residuals of a risk factors model.

    Parameters
    ----------
    residuals : ndarray or dataframe
        Ndarray or DataFrame of residuals of the risk factors model of shape n_samples x 1.

    Returns
    -------
    value : float
        Coskewness tensor of residuals of a risk factors model.

    Raises
    ------
        ValueError when the value cannot be calculated.

    """
    residuals_ = np.array(residuals, ndmin=2)

    value = cpp_residuals_coskewness_fm(residuals_)

    return value


def residuals_cokurtosis_fm(B: np.ndarray, S_f: np.ndarray, residuals: np.ndarray):
    r"""
    Calculates the cokurtosis square matrix of residuals of a risk factors model.

    Parameters
    ----------
    B : ndarray or dataframe
        The loadings matrix.

    S_f : ndarray or dataframe
        Covariance matrix of the risk factors.

    residuals : ndarray or dataframe
        Ndarray or DataFrame of residuals of the risk factors model of shape n_samples x 1.

    Returns
    -------
    value : float
        Cokurtosis square matrix of residuals of a risk factors model.

    Raises
    ------
        ValueError when the value cannot be calculated.

    """
    B_ = np.array(B, ndmin=2)
    S_f_ = np.array(S_f, ndmin=2)
    residuals_ = np.array(residuals, ndmin=2)

    value = cpp_residuals_cokurtosis_fm(B_, S_f_, residuals_)

    return value
